### PR TITLE
add native support for pv (displaying winrate and variations ingame) for some AIs

### PR DESCRIPTION
**Pull request by wonderingabout**
edit: rebased to be on top of devel

is an implementation of https://github.com/Dorus/gtp2ogs/blob/dorus/

### CODE REVIEW

File: bot.js [hunk 1]
```diff
@@ -112,83 +125,188 @@ class Bot {
...
+            }
+        }
+
+        let stopRe = config.SAISTOPRE;
+        let stop = stopRe.exec(errline);
+        if (stop && this.prevLine) {
+            let winrate = this.prevLine[3],
+                visits = stop[1],
+                playouts = stop[3],
+                nps = stop[4],
```
Comment (wonderingabout) [inline bot.js:212]: `nps` is never used, is it fine ?

File: bot.js [hunk 2]
```diff
@@ -112,83 +125,188 @@ class Bot {
...
+            }
+        }
+
+        let stopRe = config.LZSTOPRE;
+        let stop = stopRe.exec(errline);
+        if (stop && this.prevLine) {
+            let winrate = this.prevLine[3],
+                visits = stop[1],
+                playouts = stop[3],
+                nps = stop[4],
```
Comment (wonderingabout) [inline bot.js:184]: `nps` is never used, is it fine ?

**Comment by wonderingabout** [comment]
just rebased the pr to be on top of latest devel, so that minimal review is required

File: pv.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                   // .replace(this.CLPV, '') // duplicate, safe to remove ?
+                   .split(",")
+                   .map(s => s === '..' ? '..' : s[0] + num2char(this.game.state.width - char2num(s[1]) - 1))
+                   .join('');
+
+        return this.createMessage(name, pv);
+    }
+}
+
+function num2char(num) {
```
Comment (Dorus) [inline pv.js:81]: The functions and the 2 below are not just used in pv right?

We might want to export them and reused them elsewhere, or give them their own module and import them both here and the other places where they're used.

File: pv.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                  .join('');
+
+        return this.createMessage(name, pv);
+    }
+    getPvChatPG(stop) {
+        const name = stop[1];
+
+        const pv = this.pvLine[1]
+                   .replace(this.CLPV, '') 
+                   // .replace(this.CLPV, '') // duplicate, safe to remove ?
```
Comment (Dorus) [inline pv.js:72]: I think so yes. The CLPV has /g so it will replace all instances it finds.

We do need to test all 3 bots.

File: pv.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                   .split(",")
+                   .map(s => s === '..' ? '..' : s[0] + num2char(this.game.state.width - char2num(s[1]) - 1))
```
Comment (Dorus) [inline pv.js:74]: We dont have `this.game.state.width` here. We need to pass the `width` either here or in the constructor.

File: bot.js [hunk 6]
```diff
@@ -116,7 +131,17 @@ class Bot {
...
     }}}
-
+    postPvToChat(errline) {
+        this.pv.updatePvLine(errline);
+        const stop = this.pv.STOPRE.exec(errline);
+        
+        if (stop && this.pv.pvLine) {
+            const body = this.pv.getPvChat(stop);
+            const move = this.game.state.moves.length + 1;
+            this.game.sendChat(body, move, "malkovich");
```
Comment (Dorus) [inline bot.js:141]: This line will be a problem too. this.game is not available.

File: bot.js [hunk 7]
```diff
@@ -116,7 +131,17 @@ class Bot {
...
         });
     }}}
-
+    postPvToChat(errline) {
+        this.pv.updatePvLine(errline);
+        const stop = this.pv.STOPRE.exec(errline);
+        
+        if (stop && this.pv.pvLine) {
+            const body = this.pv.getPvChat(stop);
+            const move = this.game.state.moves.length + 1;
```
Comment (Dorus) [inline bot.js:140]: We also need to pass the current move.

**Comment by Dorus** [comment]
I tested the code. It's working 100% for leela zero and sai. For phoenixgo it randomly outputs extra chat messages with incomplete pv's. I'm not sure what's up with that.

**Comment by Dorus** [comment]
Wishlist:
-circle mark
-pondering suppport
-katago
-gomill
--testing

But these things can probably go into a new pr.

@wonderingabout you fix the documentation?

**Comment by roy7** [comment]
I know Dorus added circle to his copy of meta's branch I think, but meta must not have merged it over yet since the PR here doesn't include circle support?

**Comment by wonderingabout** [comment]
> I know Dorus added circle to his copy of meta's branch I think, but meta must not have merged it over yet since the PR here doesn't include circle support?

indeed, i accidentally rebased after merging @Dorus 's PR, i fixed it by manually doing the changes again in https://github.com/online-go/gtp2ogs/pull/224/commits/ccba48eaec32b60c1ceeebc89f7015dea145ebcb

**Comment by wonderingabout** [comment]
> I know Dorus added circle to his copy of meta's branch I think, but meta must not have merged it over yet since the PR here doesn't include circle support?

accidentally erased @Dorus 's commit while rebasing, we will get it back again in @Dorus 's next PR

File: docs/OPTIONS-LIST.md [hunk 8]
```diff
@@ -104,19 +104,56 @@ file exists (checked each time, can use for load-balancing)
...
+
+##### ogspv alternative weights support:
+
+You can run your AI engine **with any weight it supports**, for example 
+leela zero engine with 40b, 15b, elf-v0, elf-v1, elf-v2, minigo, etc.
+
+##### ogspv AI-specific requirements and tips:
+
+for KataGo:
+- the requirement to set `ogsChatToStderr=true` in the config.
```
Comment (Dorus) [inline docs/OPTIONS-LIST.md:153]: This is only required for pv.

File: docs/OPTIONS-LIST.md [hunk 9]
```diff
@@ -104,19 +104,56 @@ file exists (checked each time, can use for load-balancing)
...
+
+for PhoenixGo:
+- the requirement to disable pondering, you need to set `enable_background_search`
```
Comment (Dorus) [inline docs/OPTIONS-LIST.md:156]: Should also mention somewhere pondering on windows for leela 0.11 is not working.

File: pv.js [hunk 10]
```diff
@@ -8,29 +8,29 @@ class Pv {
...
                            'SAI': this.getPvChatSAI,
-                           'PG':  this.getPvChatPG,
-                           'KATA': this.getPvChatKata,
-                           'LEELA': this.getPvChatLeela
-                         }[setting];
-        this.PVRE =      { 'LZ':  (/([A-Z]\d+|pass) -> +(\d+) \(V: +(\d+.\d\d)%\) (\(LCB: +(\d+.\d\d)%\) )?\(N: +(\d+.\d\d)%\) PV:(( ([A-Z][0-9]+|pass)+)+)/),
+                           'KATAGO': this.getPvChatKata,
+                           'PHOENIXGO':  this.getPvChatPG,
+                           'LEELA': this.getPvChatLeela,
+                         }[setting.toUpperCase()];
```
Comment (Dorus) [inline pv.js:16]: Please put `.toUpperCase()` in config.js

File: config.js [hunk 11]
```diff
@@ -503,6 +512,14 @@ function testDeprecatedArgv(optimistArgv, komisFamilyNameString) {
     console.log("\n");
 }
 
+function checkUnsupportedOgspvAI(ogspv, ogsPvAIs) {
+    const upperCaseAIs = ogsPvAIs.map(e => e.toUpperCase());
+
+    if (!upperCaseAIs.includes(ogspv)) {
+        throw new `Unsupported --ogspv option ${ogspv}.`
+                  + `\nSupported options are ${AIs.join(', ')}`;
```
Comment (Dorus) [inline config.js:520]: Where does `AIs` come from?

File: config.js [hunk 12]
```diff
@@ -503,6 +512,14 @@ function testDeprecatedArgv(optimistArgv, komisFamilyNameString) {
     console.log("\n");
 }
 
+function checkUnsupportedOgspvAI(ogspv, ogsPvAIs) {
+    const upperCaseAIs = ogsPvAIs.map(e => e.toUpperCase());
+
+    if (!upperCaseAIs.includes(ogspv)) {
+        throw new `Unsupported --ogspv option ${ogspv}.`
```
Comment (Dorus) [inline config.js]: Remove `new` here.

**Comment by Dorus** [comment]
This looks good now. I tested the following bots:

kata ponder - :white_check_mark: 
kata - :white_check_mark: 
leela - :white_check_mark: 
leela zero ponder - :white_check_mark: 
leela zero - :white_check_mark: 
sai - :white_check_mark: 
sai ponder - :white_check_mark: 
phoenix go - :white_check_mark:

**Pull request merged**